### ISSUE TITLE
Sadmodules

### ISSUE THREAD

**Issue by mtschoen-unity** [ForkIssues/Tracker#888]
Sorry for the vague and silly issue title--I couldn't resist!

I'll admit that submodule workflows are particularly tricky, but here are some issues I've been having. Rather than re-open #620 I figured that I would report what I am seeing more broadly

All of this behavior occurs when `Update submodules automatically` is enabled. I would expect Fork to act like CLI otherwise, which it seems to do. For the most part here, I am talking about behaviors around checking out commits which add, move, remove, and/or change remote url's for submodules. This can be a rare occurrence for some but for me it's pretty frequent. Some of the behavior is also a little confusing when users don't know what to expect when checking out a new changeset.

I haven't had the time to work out an example repository which reproduces the issues I'm seeing, but I have seen them reliably on my and my colleague's machines.

1. It seems that any failure while updating submodules causes the entire operation to stop. I believe that this is the behavior of `git submodule update`, but I would expect that since Fork is already diverting from standard checkout behavior, it should keep trying. Maybe you could track which submods succeeded/failed and then pick up where the process left off, updating the rest in the list?
2. There is no way in the UI to trigger a `git submodule sync`, and in situations where this is necessary the normal submodule update fails. For example, if I check out a branch that switched to a fork of a particular submodule, it is likely that the original remote does not have the changeset being pointed to, and the update will fail.
3. I am still seeing the behavior from #620 where some submodules will update, but new ones fail to `--init`. It may be a symptom of `1.` where some other part of the update fails before it can get to my new submodule, but there are definitely situations where I have `Update submodules automatically` enabled, and new submodules are missing.
4. In the case of `3.`, or really any case where you have a non-initted submodule, it will appear in the list (on the lower-left under > Submodules) of submodules, and be interactable, but when you double-click it, Fork opens a new tab on the _current repository_. This is sometimes not obvious to less familiar users, who at-a-glance may not be able to tell the difference between a repository and one of its submodules. I would say that empty submodules should show up differently in the list, and should prompt you to initialize them on double-click. Or they could simply do nothing on double-click but certainly shouldn't exhibit the current behavior.

I do really appreciate that `Discard Changes` on a submodule is equivalent to `submodule update`, but this was a little hard to discover. Would it be possible to change the wording in the context menu, dialogs, etc. to be consistent with the git commands? I think it's fine to keep them both bound to `command-shift-d`, but referring to it as `Discard` is misleading IMO.

**Comment by DanPristupov** [comment]
> 1. It seems that any failure while updating submodules causes the entire operation to stop. I believe that this is the behavior of git submodule update, but I would expect that since Fork is already diverting from standard checkout behavior, it should keep trying. 

Fork just performs `git submodule update --init --recursive` after checkout. It's not a large diverse.

We should keep in mind that there's a performance trade-off caused by 'submodule update'. Performing update separately for each submodule (i.e. 'keep trying') will significantly affect the performance of checkout/pull/merge etc.

> 2. There is no way in the UI to trigger a git submodule sync

I'm not really sure what `submodule sync` is for. According to https://stackoverflow.com/questions/45678862/git-submodule-update-vs-git-submodule-sync you shouldn't need this command often, don't you? If you do, you can create a custom command for it.

> 4. In the case of 3., or really any case where you have a non-initted submodule, it will appear in the list (on the lower-left under > Submodules) of submodules, and be interactable, but when you double-click it, Fork opens a new tab on the current repository

Could you double check this please? I think we've fixed (improved) this behavior some time ago and there's a code which handles this case. 

> I do really appreciate that Discard Changes on a submodule is equivalent to submodule update

It's not. 'Discard submodule changes' performs 3 commands. 

```
git reset --hard HEAD
git clean -f

git submodule update --init --recursive --
```

The submodule **update** is available in the submodule context menu on the sidebar.

**Comment by mtschoen-unity** [comment]
>We should keep in mind that there's a performance trade-off caused by 'submodule update'. Performing update separately for each submodule (i.e. 'keep trying') will significantly affect the performance of checkout/pull/merge etc.

Fair enough. Maybe it could be a second option to enable this. I believe the way that this is handled in TortoiseGit is that a "submodule update" is always a series of individual updates of whatever submodules you have selected. I think Fork could handle this similarly, and that way if one command fails, the rest of the sequence can carry on. I'm not sure if there are performance implications to a single command vs. one command per submodule but this seems to work well on TortoiseGit.

> I'm not really sure what submodule sync is for

It updates the remote configuration of submodules to match the urls in `.gitmodules`. For example, if a submodule was added to the project, and changes were made to that submodule but the person who made them does not have push access, they will have to fork the repo and set up a new remote URL. The only way for them to commit this change is to change the `.gitmodules` file with the new URL. If I'm updating to this branch, and I've already checked out the repo and have the submodule initialized using the old URL, when I try to update that submodule, I'll see an error that reads `Unadvertised commit: blahblah` because it is still trying to fetch from the old URL. If I know what all of this means, I'll then run `git submodule sync` to apply the new URL from .gitmodules to my local checkout of that submodule, and then `submodule update` again to get it to fetch the new changes. The sync command is pretty quick and while in some rare cases where users don't want their remote config overridden, it doesn't actually change anything. I think it would be nice to add an option like `update submodules automatically` to do a sync before updating to prevent issues.

> you shouldn't need this command often, don't you? If you do, you can create a custom command for it.

As I said, most people don't need this often but I do a lot of checking out other peoples' branches, and sometimes that means they've made a change to a submodule URL. I won't know about this change unless I'm looking out for it, and it might be nice to include this as an option to always sync when checking out branches. It's usually pretty quick. To be clear, the issue isn't that it happens frequently, but that it's confusing when it needs to happen and results in an unadvertised commit error down the line.

> Could you double check this please? I think we've fixed (improved) this behavior some time ago and there's a code which handles this case.

This was the case the same day I posted this issue, and I'm pretty sure I'm on the latest version. I will update this issue if I see it happen again.

> It's not. 'Discard submodule changes' performs 3 commands.

Ah. Good to know. The issue with using `submodule update` from the sidebar is that I need to update each submodule individually. I can multi-select submodules in the change list and "discard changes" when I want to do an update. Can we add a "submodule update" item to the context menu for the list of all submodules (and folders?), as well as the one that comes up for individual submodules.

**Comment by mtschoen-unity** [comment]
Unfortunately I've just encountered this issue (checkout branch but fail to init submodules) again on the latest Fork (1.0.94).

Here's my log for today which does seem to contain some relevant errors. Hope this helps!
[Fork.log](https://github.com/ForkIssues/Tracker/files/4743082/Fork.log)

**Comment by mtschoen-unity** [comment]
Hey! It's been an eternity, and I've mostly been working on monorepos these past few years, but I finally came back to revisit some older repos with submodules and I was pleasantly surprised that checking out different branches that add/remove submods and generally working with submods in fork is a much more pleasant experience. I'm not sure if all of my specific feedback from above has been addressed, but it's been long enough that I think we can close this out. I'll open new tickets if I run into any issues going forward.

Thanks for making Fork better! 🤜 🤛
